src/ctx.cpp

```diff
@@ -506,8 +506,10 @@ void FunctionEmitContext::EndIf() {
 
     // 'uniform' ifs don't change the mask so we only need to restore the
     // mask going into the if for 'varying' if statements
-    if (ci->IsUniform() || bblock == nullptr)
```

Comment:
Would making `ci` a `unique_ptr` be a better / more elegant solution? I.e. declare it as `std::unique_ptr<CFInfo> ci(popCFState());`.

What's your opinion?

Reply:
I think the main advantage would be in the case the assertion is tripped, the destructor will still be called.

Reply:
Actually, I think this may be an issue for the path where the predicate is false, since the pointer should remain undeleted according to the current implementation. With unique_ptr, it will go out of scope later and be destructed. Unique_ptr assumes ownership.

Reply:
I don't understand the last comment. Both explicit delete and `uniquie_ptr` will not change the current logic in this function.

Reply:
I mean that as it is now, if `! (ci->IsUniform() || bblock == nullptr)`, there's no call to delete.

Reply:
Ah never mind, I was looking at the left diff.

Reply:
I am not sure about `unique_ptr`.
The problem here is that `popCFState` doesn't really create or allocate `CFInfo` object. The lifetime of `CFInfo` objects is the following (let's consider `if` only):
- Created via allocation on heap in `CFInfo::GetIf`: https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L145
- `CFInfo::GetIf` returns a pointer to a newly allocated object that stored in `controlFlowInfo` vector in two places:
https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L495
and
https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L499
- The pointer is stored there until being poped in `popCFState` https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L3771
- Then this pointer is returned by `popCFState` and accessed, e.g., in `EndIf` here.
So, we may bind lifetime of `CFInfo` object inside `CFInfo::GetIf` (and similar functions) to stack allocated `unique_ptr` but not the whole.

Reply:
Bearing that in mind, I have fixed that to use `unique_ptr` instead explicit `delete ci` that is better especially when multiple exit points exist.

Reply:
`unique_ptr` doesn't need to manage the whole lifetime from creation to destruction, managing it from popping to destruction is ok from philosophical point of view.